### PR TITLE
Add improved type errors with nested context

### LINKED ISSUE #16: Add more context to error messages
We can usually add some additional context such as the name of the attribute. Additionally, collection types could implement an alternative interface that allows them to highlight specifically which member failed and why.

### PR DESCRIPTION
**Pull request by segiddins**
Closes #16

### CODE REVIEW

File: lib/literal.rb [hunk 1]
```diff
@@ -25,15 +25,15 @@ def self.Enum(type)
...
 	end
 
-	def self.check(actual, expected, &context)
-		if expected === actual
+	def self.check(actual, expected, &)
+		raise ArgumentError.new("Cannot check type without a block") unless block_given?
+
+		if expected.respond_to?(:check)
+			expected.check(actual, &)
+		elsif expected === actual
```
Comment (joeldrapper) [inline lib/literal.rb:33]: Should we run `===` first and only do `check` on failure? All failures will take twice as long. 🤔

**Comment by segiddins** [comment]
@joeldrapper this should be ready for another look now!

**Comment by segiddins** [comment]
It looks like quickdraw is trying to run its own tests, instead of the tests in this repo

**Comment by stevegeek** [comment]
@segiddins maybe related to this ? https://github.com/joeldrapper/quickdraw/issues/38

**Comment by joeldrapper** [comment]
Oh yeah, it's the way GitHub bundles gems. We can ignore it. I think there's probably a way to exclude that folder from the Quickdraw glob but haven't looked into it. It only happens when you pull the gem from GitHub.

**Comment by joeldrapper** [comment]
Sorry for the delay on reviewing this @segiddins. I will get to it soon!

**Comment by joeldrapper** [comment]
This is a fantastic PR @segiddins and I know it was a ton of work. Thank you for digging into this. 🙌

**Comment by stevegeek** [comment]
Amazing thanks @segiddins !

**Pull request merged**